### PR TITLE
Can restore deleted objective

### PR DESCRIPTION
**Pull request by denis1stomin**
The functionality is not fully implemented.
TODO: add restore button or link.

But if someone of you is interested to change deleted objective style you're welcome :)))
while I'm finishing restore button.

### CODE REVIEW

**Comment by preigile** [comment]
Чуть подправила стили. Для удалённых кирезалтов не будешь то же самое делать? И ещё, наверное, кнопки надо спрятать, вместо них Рестор нарисовать, чтобы не лайкать/удалять/копировать

**Comment by denis1stomin** [comment]
Yep, will add restore "button".
I think it doesn't make sense to restore KRs but can be wrong

**Comment by denis1stomin** [comment]
Almost works as expected.
Need to fix:
1. styles - I cannot make the obj card to be flexible and fit the window
2. fake objective is broken
3. KRs are not restored since the objective is re-created
3. Removed objective become last one in the list (may be Ok for some time)
4. When copying objective - removed one is lost

**Comment by denis1stomin** [comment]
Fixed:
2. fake objective is broken

TODO: refactoring - make objective as separate component

**Comment by preigile** [comment]
error in console: `[Vue warn]: Property or method "canChangeOkr" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.`

**Comment by preigile** [comment]
I've changed some styles

**Comment by skalitin** [comment]
I would not use additional collection to store deleted objectives - this way it is pretty hard to maintain their order. I would add new property `isDeleted` to the existing objectives, and on 'Delete' click either:
- set this property to true, do not delete objective from local store but send delete request to OneNote. To restore objective we need to send 'create objective' request.
- or even set this property to true and send change objective request to OneNote, which will mark objective as deleted by for example adding __deleted__ prefix to the page name. To restore objective we need to send 'change objective' request. Benefits - it is very easy to keep objectives order and key results. Actually it would be much better to use built-in OneNote Deleted Pages, but they are not available through API. 

What do you think?

**Comment by skalitin** [comment]
IMHO current 'Restore Objective' style is too aggressive, it should not be the focus of user's attention is it is not, at least for me. I would expect it to look like this:
image

**Review by skalitin — COMMENTED** [review]
Please see my comments below

**Comment by denis1stomin** [comment]
> error in console: [Vue warn]: Property or method "canChangeOkr" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.

Can't see it since I use production build and ngnix host

**Comment by denis1stomin** [comment]
> What do you think?

`isDeleted` property was my first approach to implement the functionality.
And I agree it is much easier to implement. But really I don't like to mix properties and meta-properties (may be it is not good rule for javascript :smiley:) so I decided to use additional collection.
Then I met the issue with sorting since delete objective jumps to the end of the list.
I have been thinking about it and decided that it is Ok since in future we want to provide ability for the user to change objectives order. Therefore we will introduce some property like `orderIndex` or `createdAt` and will sort the list using this property anyway. So it looks Ok to not have the order right now if it will be introduced later.
The next problem is to restore key results along with its objective.
There was a lot thoughts similar to yours and the final one was to introduce `Trash` book section. So we simply will move deleted objectives into the `trash section` and provide possibility to restore an objective from the trash or to `empty trash`. The functionality should be based on possibility to group objectives (using different notebook sections) which I believe is very important functionality in future.
So I'd like to push current changes and improve it little bit later.
At least we really can restore objective currently :smiley: may be not the best way I agree.

**Comment by denis1stomin** [comment]
> current 'Restore Objective' style is too aggressive

Please take a look at current style.
The main idea of adding transparent layer is to show some color information from a deleted objective card. I mean that different objectives will have different colors in future.
Also dark style helps objective to jump out of the user focus.
(Agree that red was not the best choice. Now the color is grey.)

**Comment by denis1stomin** [comment]
Not fixed yet:
- KRs are not restored since the objective is re-created
- Removed objective become last one in the list (may be Ok for some time)
- When copying objective - removed one is lost
- refactoring - make objective as separate component

I think it's better to make the changes separately from this change.
What do you think guys:grey_question:  :)

**Comment by denis1stomin** [comment]
Stupidly fixed:
- KRs are not restored since the objective is re-created
Almost works but I'm getting the error `Cannot read property 'id' of null` which I will fix soon

**Comment by skalitin** [comment]
> Please take a look at current style.

I saw it (grey one), and for me it is very aggressive :)

> The main idea of adding transparent layer is to show some color information from a deleted objective card. I mean that different objectives will have different colors in future.

Is this color information important for deleted objective? I don't think so. Let me be clear here - we are talking about restoring objectives which were deleted *a moment ago* - not an hour ago, not yesterday. We are not implementing Recycle Bin/Trash (yet). Btw I am sure that Trash is not required here - objectives are too simple things by its nature. And for such objectives it is enough to show only basic information - maybe even only objective's title (but title and KRs is better because it clearly shows that KRs will be restored too).

**Comment by skalitin** [comment]
> And I agree it is much easier to implement. But really I don't like to mix properties and meta-properties (may be it is not good rule for javascript 😃) so I decided to use additional collection.

Sorry about it, but I don't like two collections approach a lot. :) It is over-complicated without real need. The `orderIndex` property will not help, or at least will add even more complexity - how are you going to have consistent indexes in the first collection without taking into account indexes from the second collection?

**Comment by denis1stomin** [comment]
Fixed:
- Removed objective become last one in the list (may be Ok for some time)
- When copying objective - removed one is lost

BTW Added KRs handling when copy an objective.

**Comment by preigile** [comment]
Иногда после удаления улетает вниз, иногда нет. 
Кирезалты не восстанавливаются.

1) Создала objective
2) Удалила Objective
3) Добавила к нему keyresulr
4) Снова удалила
5) Восстановила и обновила страничку
6) Вижу 2 одинаковых objectives без keyresult

**Comment by preigile** [comment]
error in console: `[Vue warn]: Property or method "canChangeOkr" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.`
Всё ещё воспроизводится

**Comment by preigile** [comment]
А ещё карточки начали скакать, например, когда карточка находится не сверху, а я добавляю ей keyresult, то она бежит наверх

**Comment by skalitin** [comment]
> Property or method "canChangeOkr" is not defined on the instance but referenced during render.

Fixed

**Comment by skalitin** [comment]
> 2. Удалила Objective
> 3. Добавила к нему keyresulr

Это как?

**Comment by preigile** [comment]
> Удалила Objective
> Добавила к нему keyresulr
> Это как?

Удалила -> Восстановила -> Добавила kr

File: webui/src/store/modules/okr.js [hunk 1]
```diff
@@ -138,60 +180,35 @@ export default {
...
 
         CREATE_OBJECTIVE({state, commit}, objective) {
-            // change local objectives list
-            let changedList = state.objectives;
-            changedList.push(objective);
-            commit('OBJECTIVES_COMPLETE', changedList);
+            // We need to set unique objective id before committing mutation,
+            // otherwise Vue's list rendering engine (v-for statement) will reuse DOM elements
+            // and display several items with the same text for newly added objectives.
+            objective.id = okrSvc.createId();
```
Comment (denis1stomin) [inline webui/src/store/modules/okr.js]: May be we can make this something like `"temp-" + okrSvc.createId()` and hide `remove` button when `ID` starts with `temp`?

**Comment by denis1stomin** [comment]
> Удалила -> Восстановила -> Добавила kr

Definitely you should make it slower ~~baby~~ 😄

**Review by preigile — APPROVED** [review]

**Pull request merged**